crates/assists/src/handlers/infer_function_return_type.rs

```diff
@@ -0,0 +1,237 @@
+use hir::HirDisplay;
+use syntax::{ast, AstNode, SyntaxToken, TextSize};
+use test_utils::mark;
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: infer_function_return_type
+//
+// Adds the return type to a function or closure inferred from its tail expression if it doesn't have a return
+// type specified.
+//
+// ```
+// fn foo() { 4<|>2i32 }
+// ```
+// ->
+// ```
+// fn foo() -> i32 { 42i32 }
+// ```
+pub(crate) fn infer_function_return_type(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let expr = ctx.find_node_at_offset::<ast::Expr>()?;
+    let (tail_expr, insert_pos, wrap_expr) = extract_tail(expr)?;
+    let module = ctx.sema.scope(tail_expr.syntax()).module()?;
+    let ty = ctx.sema.type_of_expr(&tail_expr).filter(|ty| !ty.is_unit())?;
+    let ty = ty.display_source_code(ctx.db(), module.into()).ok()?;
+
+    acc.add(
+        AssistId("infer_function_return_type", AssistKind::RefactorRewrite),
+        "Add this function's return type",
+        tail_expr.syntax().text_range(),
+        |builder| {
+            let insert_pos = insert_pos.text_range().end() + TextSize::from(1);
+            builder.insert(insert_pos, &format!("-> {} ", ty));
+            if wrap_expr {
+                mark::hit!(wrap_closure_non_block_expr);
+                // `|x| x` becomes `|x| -> T x` which is invalid, so wrap it in a block
+                builder.replace(tail_expr.syntax().text_range(), &format!("{{{}}}", tail_expr));
+            }
+        },
+    )
+}
+
+fn extract_tail(expr: ast::Expr) -> Option<(ast::Expr, SyntaxToken, bool)> {
+    let (ret_ty, tail_expr, insert_pos, wrap_expr) =
+        if let Some(closure) = expr.syntax().ancestors().find_map(ast::ClosureExpr::cast) {
+            let (tail_expr, wrap_expr) = match closure.body()? {
```

Comment:
hey--commenting out of curiosity but would this always infer function types for lambdas? Is that something that's wanted?

Reply:
It would give you the option to have this assist explicitly annotate the closure return type if you are in the last expression of the closure. This is an action the user would have to trigger. So the user has to trigger the assist to have it annotate the closure with its explicit return type.

The assist name is poorly chosen I think but I couldn't come up with a better name 😅